src/LinearSolvers/preconditioners.jl

```diff
@@ -151,6 +196,49 @@ function BlockJacobiPreconditioner(J::SparseMatrixCSC; nblocks=-1, device=CPU())
     return BlockJacobiPreconditioner(J, npartitions, device)
 end
 BlockJacobiPreconditioner(J::CUSPARSE.CuSparseMatrixCSR; options...) = BlockJacobiPreconditioner(SparseMatrixCSC(J); options...)
+Base.eltype(::BlockJacobiPreconditioner{AT,GAT,VI,GVI,MT,GMT,MI,GMI,SMT}) where {AT,GAT,VI,GVI,MT,GMT,MI,GMI,SMT} = Float64
+
+@kernel function multiply_blocks_gpu!(y, b, p_len, rp_len, part, blocks)
+    i = @index(Global, Linear)
+    len = p_len[i]
+    rlen = rp_len[i]
+    idxA = -1
+    idxB = -1
+    accum = 0.0
+    # homemade matrix multiply. This is probably not a good idea.
+    @inbounds for j=1:rlen
+        idxA = part[j, i]
+        accum = 0.0
+        @inbounds for k=1:len
+            idxB = part[k, i]
+            accum = accum + blocks[j, k, i]*b[idxB]
+        end
+        y[idxA] = accum
+    end
+end
+
+@inline function mul!(y, C::BlockJacobiPreconditioner, b::Vector{Float64})
+    n = size(b, 1)
+    y .= 0.0
+    for i=1:C.nblocks
+        rlen = C.lpartitions[i]
+        part = C.partitions[1:rlen, i]
+        blck = C.blocks[1:rlen, 1:rlen, i]
+        for j=1:C.rest_size[i]
+            idx = part[j]
+            y[idx] += dot(blck[j, :], b[part])
+        end
+    end
+end
+
+@inline function mul!(y, C::BlockJacobiPreconditioner, b::CuVector{Float64})
+    n = size(b, 1)
```

Comment:
Yea, I still don't understand, but I tend to use `fill!(y, zero(Float64))`. `fill!` for compiling faster and `zero(Float64)` for avoiding type weirdness (since using Enzyme).

Reply:
That makes sense indeed.